relay-general/src/pii/attachments.rs

```diff
@@ -71,43 +65,283 @@ fn apply_regex_to_bytes(
+            }
+            Redaction::Replace(ref replace) => {
+                self.swap_content(replace.text.as_str(), PADDING);
             }
         }
-        Redaction::Replace(ref replace) => {
-            for (start, end) in matches {
-                replace_bytes_padded(
-                    replace.text.as_bytes(),
-                    &mut data[start..end],
-                    DEFAULT_PADDING,
-                );
+    }
+}
+
+impl StringMods for WStr {
+    type Error = Utf16Error;
+
+    fn fill_content(&mut self, fill_char: char) {
+        let size = std::mem::size_of::<u16>();
+
+        let mut buf = [0u16; 1];
+        let fill_u16 = fill_char.encode_utf16(&mut buf[..]); // this panics for us
+        let fill_buf = fill_u16[0].to_le_bytes();
+
+        let chunks = self.as_bytes_mut().chunks_exact_mut(size);
+        for chunk in chunks {
+            chunk.copy_from_slice(&fill_buf);
+        }
+    }
+
+    fn swap_content(&mut self, replacement: &str, padding: char) {
+        let size = std::mem::size_of::<u16>();
+        let len = self.len();
+
+        let mut buf = [0u16; 1];
+        let fill_u16 = padding.encode_utf16(&mut buf[..]); // this panics for us.
+        let fill_buf = fill_u16[0].to_le_bytes();
+
+        let mut offset = 0;
+        for code in replacement.encode_utf16() {
+            let char_len = if 0xD800 & code == 0xD800 {
+                size * 2 // leading surrogate
+            } else {
+                size
+            };
+            if (len - offset) < char_len {
+                break; // Not enough space for this char
             }
+            let target = &mut self.as_bytes_mut()[offset..offset + size];
+            target.copy_from_slice(&code.to_le_bytes());
+            offset += size;
+        }
+
+        let remainder_bytes = &mut self.as_bytes_mut()[offset..];
+        let chunks = remainder_bytes.chunks_exact_mut(size);
+        for chunk in chunks {
+            chunk.copy_from_slice(&fill_buf);
         }
     }
+}
 
-    true
+impl StringMods for [u8] {
+    type Error = Utf8Error;
+
+    fn fill_content(&mut self, fill_char: char) {
+        let mut buf = [0u8; 1];
+        fill_char.encode_utf8(&mut buf[..]); // this panics for us
+        for byte in self {
+            *byte = buf[0];
+        }
+    }
+
+    fn swap_content(&mut self, replacement: &str, padding: char) {
+        let mut buf = [0u8; 1];
+        padding.encode_utf8(&mut buf[..]); // this panics for us
+
+        let cutoff = replacement.len().min(self.len());
+        let (left, right) = self.split_at_mut(cutoff);
+        left.copy_from_slice(&replacement.as_bytes()[..cutoff]);
+
+        for byte in right {
+            *byte = buf[0];
+        }
+    }
+}
+
+struct MutSegmentIter<'a> {
+    data: &'a mut [u8],
+    decoder: Box<dyn RawDecoder>,
+    offset: usize,
+}
+
+impl<'a> MutSegmentIter<'a> {
+    fn new(data: &'a mut [u8], encoding: impl Encoding) -> Self {
```

Comment:
_nit_: Please make this a type parameter. There's no benefit to this, but it matches our code style.

Reply:
do you mean `fn new<E: Encoding>(..., encoding: E)`?

Reply:
Yes, exactly.